jctools-core/src/main/java/org/jctools/queues/MpmcTransferArrayQueue.java

```diff
@@ -0,0 +1,613 @@
+/*
+ * Copyright 2015 dorkbox, llc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jctools.queues;
+
+import java.util.concurrent.LinkedTransferQueue;
+import java.util.concurrent.ThreadLocalRandom;
+
+import org.jctools.util.UnsafeAccess;
+
+abstract class CI1<E> {
+    public int type = 0;
+    public E item = null;
+}
+
+abstract class P0<E> extends CI1<E> {
+    volatile long y0, y1, y2, y4, y5, y6 = 7L;
+}
+
+abstract class HI0<E> extends P0<E> {
+    public Thread thread;
+}
+
+class Node<E> extends HI0<E> {
+    private static final long TYPE;
+    private static final long ITEM;
+    private static final long THREAD;
+
+    static {
+        try {
+            TYPE = UnsafeAccess.UNSAFE.objectFieldOffset(Node.class.getField("type"));
```

Comment:
Why are these fields public? There is the `getDeclaredField()` to gain access to private fields.